src/uct/cuda/cuda_ipc/cuda_ipc_ep.c

```diff
@@ -31,16 +35,155 @@ UCS_CLASS_DEFINE_NEW_FUNC(uct_cuda_ipc_ep_t, uct_ep_t, uct_iface_t*,
+            return UCS_ERR_IO_ERROR;
+        }
+
+        *mapped_rem_addr = (void *) ((uintptr_t) mapped_addr + offset);
+    }
+    return UCS_OK;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+uct_cuda_ipc_post_cuda_async_copy(uct_ep_h tl_ep, uint64_t remote_addr,
+                                  const uct_iov_t *iov, uct_rkey_t rkey,
+                                  uct_completion_t *comp, int direction)
+{
+    uct_cuda_ipc_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_cuda_ipc_iface_t);
+    uct_cuda_ipc_ep_t *ep       = ucs_derived_of(tl_ep, uct_cuda_ipc_ep_t);
+    uct_cuda_ipc_key_t *key     = (uct_cuda_ipc_key_t *) rkey;
+    void *mapped_rem_addr       = NULL;
+    uct_cuda_ipc_event_desc_t *cuda_ipc_event;
+    ucs_queue_head_t *outstanding_queue;
+    ucs_status_t status;
+    CUdeviceptr dst, src;
+    CUdevice cu_device;
+    CUstream stream;
+
+    if (0 == iov[0].length) {
+        ucs_trace_data("Zero length request: skip it");
+        return UCS_OK;
+    }
+
+    GET_CUDA_DEVICE(cu_device);
+
+    status = uct_cuda_ipc_get_mapped_addr(ep, key, cu_device, remote_addr,
+                                          &mapped_rem_addr, iov[0].buffer);
+    if (UCS_OK != status) {
+        return status;
+    }
+
+    if (0 == iface->streams_initialized) {
+        status = uct_cuda_ipc_iface_init_streams(iface);
+        if (UCS_OK != status) {
+            return status;
+        }
+    }
+
+    stream            = iface->stream_d2d[key->dev_num];
+    outstanding_queue = &iface->outstanding_d2d_event_q;
+    cuda_ipc_event    = ucs_mpool_get(&iface->event_desc);
+
+    if (ucs_unlikely(cuda_ipc_event == NULL)) {
+        ucs_error("Failed to allocate cuda_ipc event object");
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    dst = (CUdeviceptr)
+        ((direction == UCT_CUDA_IPC_PUT) ? mapped_rem_addr : iov[0].buffer);
+    src = (CUdeviceptr)
+        ((direction == UCT_CUDA_IPC_PUT) ? iov[0].buffer : mapped_rem_addr);
+
+    status = UCT_CUDADRV_FUNC(cuMemcpyDtoDAsync(dst, src, iov[0].length, stream));
+    if (UCS_OK != status) {
+        return status;
+    }
+
+    status = UCT_CUDADRV_FUNC(cuEventRecord(cuda_ipc_event->event, stream));
+    if (UCS_OK != status) {
+        return status;
+    }
+
+    ucs_queue_push(outstanding_queue, &cuda_ipc_event->queue);
+    cuda_ipc_event->comp = comp;
+    ucs_trace("cuMemcpyDtoDAsync issued :%p dst:%p, src:%p  len:%ld",
+             cuda_ipc_event, (void *) dst, (void *) src, iov[0].length);
+    return UCS_INPROGRESS;
+}
+
 ucs_status_t uct_cuda_ipc_ep_get_zcopy(uct_ep_h tl_ep, const uct_iov_t *iov, size_t iovcnt,
                                        uint64_t remote_addr, uct_rkey_t rkey,
                                        uct_completion_t *comp)
 {
-    return UCS_OK;
+    ucs_status_t status;
+
+    status = uct_cuda_ipc_post_cuda_async_copy(tl_ep, remote_addr, iov,
+                                               rkey, comp, UCT_CUDA_IPC_GET);
```

Comment:
in case of error need to return status here, to avoid counting/printing info about failed op